shared/cancel/canceller.go

```diff
@@ -0,0 +1,19 @@
+package cancel
+
+import "context"
+
+// Canceller is a simple wrapper for a cancellable context which makes the associated context.CancelFunc more easily
+// accessible.
+type Canceller struct {
+	context.Context
+	Close func()
```

Comment:
Shall we call this `Cancel` actually, as that keeps it inline with the `HTTPRequestCanceller.Cancel()` function and the context.Context convention of using the `cancel` function var name.